### ISSUE TITLE
Frozen user.slice + child processes after resuming from suspend to RAM / standby

### ISSUE THREAD

**Issue by T-X** [systemd/systemd#33795]
### systemd version the issue has been seen with

256.2-1

### Used distribution

Debian Sid

### Linux kernel version used

6.9.8-amd64 (6.9.8-1)

### CPU architectures issue was seen on

x86_64

### Expected behaviour you didn't see

After closing the lid of my laptop and opening it again, so after resuming from standby / suspend to RAM, all processes should continue running as before.

### Unexpected behaviour you saw

After opening the laptop's lid again on Xmonad / X11 I could no more change or interact with applications, windows or workspaces on X11. Only moving the mouse would work and result in noticeable responses.

After switching to a VT and logging in in there I could see in "$ systemctl status user.slice" that it was still "Active: active (freezing)". And any childs had: "Active: active (freezing-by-parent)".

"sudo init 3; sudo init 5" from the VT would bring me back to a new, working session, while the processes from the other session would still exist in the background.

"sudo systemctl thaw user.slice" from a VT would seemingly fix the problem for me.

Issue started to appear with the update I did from 255.5-1 to 256~rc3-7 on 2024-06-05, I think. Back then happened nearly always. However now it happens a lot less often and is more difficult to reproduce.

### Steps to reproduce the problem

Close the laptop lid (my laptop: Lenovo Thinkpad T14s AMD Gen3)  with systemd putting it into suspend-to-RAM. Then some minutes later (< 15 minutes) open the lid again.

Happens for me about once or twice a week now.

Not quite sure yet if NFS/KVM is required like #33083 said (I did indeed have both running when it froze for me two days ago, I think, which the attached logs are from). And I indeed have an AMD/ATI Radeon GPU with the OpenSource amdgpu driver (so not sure if related to #33462?).

### Additional program output to the terminal or log subsystem illustrating the issue

https://gist.github.com/T-X/8a4b3239f6b93099b81f5be57a879c3c

Edit: added laptop model, Lenovo Thinkpad T14s AMD Gen3

**Comment by YHNdnzj** [comment]
> "sudo systemctl thaw user.slice" from a VT would seemingly fix the problem for me.

In that case, #33093 should make things more robust for you, probably to the extent that it isn't visible at all. But, as mentioned in #33083, the actual problem lies in the kernel, as it should never lock itself up like this.

**Comment by YHNdnzj** [comment]
I'll close this as duplicate of #33269. I don't think there's anything left for systemd to resolve/workaround really.

**Comment by T-X** [comment]
@YHNdnzj ah, thanks, yes that sounds plausible and now I see too, that I have this "Failed to freeze unit 'user.slice': Connection timed out" in journalctl.

Any specific systemd version I should watch out for that should have this fix / timeout increase from 10 to 60 seconds?

**Comment by T-X** [comment]
Regarding kernel hangs, yes, also since about 1-2 months I also have weird hangs when running `zsh -c "unlimit; screen` (and since I found that culprit removed the "unlimit" line from my .zshrc). Reported this here with more details, including strace and perf measurements: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1076380

And also since about 1-2 months it takes running qemu-system-x86_64 with KVM on an OpenWrt x86-64 image several minutes while hogging one CPU core at 100% before I see the first kernel messages from the VM guest on standard out (haven't reported that one yet).

Just mentioning these in case they might be related somehow... I also don't think I did any major changes to my Debian system other than running `apt-get update && apt-get dist-upgrade` in the last 2 months.

**Comment by T-X** [comment]
Hi, I'm still seeing this issue on Debian Sid with Linux 6.11.10 and systemd 257~rc3-1, both as provided by Debian, if I have a KVM instance running. Without the KVM instance waking up works without any issues.

The KVM instance is running on Debian Sid with the same Linux and systemd versions. It is started like this:

```
$ qemu-system-x86_64 \
-enable-kvm -nodefaults -smp 1 -m 1024M -serial mon:stdio -nographic \
-kernel /home/linus/mesh-node/boot/vmlinuz-6.11.10-amd64 \
-initrd /home/linus/mesh-node/boot/initrd.img-6.11.10-amd64 -append \
nodeid=1 slub_debug console=ttyS0 uts=1 rootflags=ro ro net.ifnames=1 ip=192.168.23.2::192.168.23.1:255.255.255.0:mesh-node:ens2:off root=/dev/nfs nfsroot=192.168.23.1:/home/linus/mesh-node \
-device e1000,netdev=net0,mac=02:04:64:a4:39:b1 \
-netdev tap,id=net0,ifname=vmtap0,script=/home/linus/dev-priv/vm/ens2-ifup.sh \
-device e1000,netdev=net1,mac=02:04:64:a4:39:c1 \
-netdev tap,id=net1,ifname=vmtap1,script=no -device virtio-net-pci,netdev=net2,mac=02:04:64:a4:39:d1 \
-netdev tap,id=net2,ifname=vmtap2,script=no \
-device e1000,netdev=net3,mac=02:04:64:a4:39:e1 \
-netdev tap,id=net3,ifname=vmtap3,script=no \
-device virtio-net-pci,netdev=net4,mac=02:04:64:a4:39:f1 \
-netdev tap,id=net4,ifname=vmtap4,script=no \
-fsdev local,security_model=passthrough,id=fsdev-root,path=/home/linus/mesh-node \
-device virtio-9p-pci,id=fs-root,fsdev=fsdev-root,mount_tag=virtfs_root
```

Notably, without the KVM instance then a "sudo systemctl suspend" puts this laptop into standby nearly immediately. However with the KVM instance this command seems to wait for about 60 seconds and then the laptop starts to sleep.

Logs from systemd-suspend.service look as follows, first half without a KVM instance (which works fine), second half with a KVM instance running (which keeps the user.slice frozen after returning from standby):

```
$ sudo journalctl -b-1 -u systemd-suspend.service
Dec 04 14:26:08 linus-lptp systemd[1]: Starting systemd-suspend.service - System Suspend...
Dec 04 14:26:08 linus-lptp systemd-sleep[5252]: Successfully froze unit 'user.slice'.
Dec 04 14:26:08 linus-lptp systemd-sleep[5252]: Performing sleep operation 'suspend'...
Dec 04 14:26:20 linus-lptp systemd-sleep[5252]: System returned from sleep operation 'suspend'.
Dec 04 14:26:20 linus-lptp systemd-sleep[5252]: Successfully thawed unit 'user.slice'.
Dec 04 14:26:20 linus-lptp systemd[1]: systemd-suspend.service: Deactivated successfully.
Dec 04 14:26:20 linus-lptp systemd[1]: Finished systemd-suspend.service - System Suspend.
Dec 04 14:28:15 linus-lptp systemd[1]: Starting systemd-suspend.service - System Suspend...
Dec 04 14:29:15 linus-lptp systemd-sleep[5973]: Failed to freeze unit 'user.slice': Connection timed out
Dec 04 14:29:15 linus-lptp systemd-sleep[5973]: Performing sleep operation 'suspend'...
Dec 04 14:29:38 linus-lptp systemd-sleep[5973]: System returned from sleep operation 'suspend'.
Dec 04 14:29:38 linus-lptp systemd-sleep[5973]: Successfully thawed unit 'user.slice'.
Dec 04 14:29:38 linus-lptp systemd[1]: systemd-suspend.service: Deactivated successfully.
Dec 04 14:29:38 linus-lptp systemd[1]: Finished systemd-suspend.service - System Suspend.
```

This is reproduceable for me every time.

**Comment by bluca** [comment]
That is a separate kernel bug, fixed in 6.13, see: https://github.com/systemd/systemd/issues/33083
